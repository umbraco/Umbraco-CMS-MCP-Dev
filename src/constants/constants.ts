export const BLANK_UUID = "00000000-0000-0000-0000-000000000000";
export const ROOT_DOCUMENT_TYPE_ID = "a95360e8-ff04-40b1-8f46-7aa4b5983096";
export const CONTENT_DOCUMENT_TYPE_ID = "b871f83c-2395-4894-be0f-5422c1a71e48";
export const Default_Memeber_TYPE_ID = "d59be02f-1df9-4228-aa1e-01917d806cda";
export const TextString_DATA_TYPE_ID = "0cc0eba1-9960-42c9-bf9b-60e150b429ae";
export const IMAGE_MEDIA_TYPE_ID = "cc07b313-0843-4aa8-bbda-871c8da728c8";
export const FOLDER_MEDIA_TYPE_ID = "f38bd2d7-65d0-48e6-95dc-87ce06ec2d3d";
export const EXAMPLE_IMAGE_PATH =
  "/src/umb-management-api/tools/temporary-file/__tests__/helpers/example.jpg";
