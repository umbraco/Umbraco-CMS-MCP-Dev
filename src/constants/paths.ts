import { join } from 'path';

// Use absolute path from current working directory
export const RESOURCES_DIR = join(process.cwd(), 'build', 'resources').replace(/\\/g, '/');