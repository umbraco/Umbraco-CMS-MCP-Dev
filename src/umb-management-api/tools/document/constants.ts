export const UmbracoDocumentPermissions = {
    Create: "Umb.Document.Create",
    Read: "Umb.Document.Read",
    Update: "Umb.Document.Update",
    Delete: "Umb.Document.Delete",
    Publish: "Umb.Document.Publish",
    Unpublish: "Umb.Document.Unpublish",
    Move: "Umb.Document.Move",
    Sort: "Umb.Document.Sort",
    Duplicate: "Umb.Document.Duplicate",
    PublicAccess: "Umb.Document.PublicAccess",
    CultureAndHostnames: "Umb.Document.CultureAndHostnames",
    CreateBlueprint: "Umb.Document.CreateBlueprint",
}