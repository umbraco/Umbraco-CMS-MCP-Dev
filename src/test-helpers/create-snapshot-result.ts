import { BLANK_UUID } from "@/constants/constants.js";
import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

export function createSnapshotResult(result: any, idToReplace?: string) {
  if (!result?.content) {
    return result;
  }

  function normalizeItem(i: any) {
    const item = { ...i, id: BLANK_UUID };
    if (item.parent) {
      item.parent = { ...item.parent, id: BLANK_UUID };
    }
    if (item.createDate) {
      item.createDate = "NORMALIZED_DATE";
    }
    if (item.publishDate) {
      item.publishDate = "NORMALIZED_DATE";
    }
    if (item.updateDate) {
      item.updateDate = "NORMALIZED_DATE";
    }
    return item;
  }

  return {
    ...result,
    content: result.content.map((item: any) => {
      if (item.type === "text") {
        if (idToReplace) {
          // For single item responses
          let text = item.text.replace(idToReplace, BLANK_UUID);
          try {
            const parsed = JSON.parse(text);
            if (parsed.createDate) {
              parsed.createDate = "NORMALIZED_DATE";
            }
            if (parsed.availableUntil) {
              parsed.availableUntil = "NORMALIZED_DATE";
            }
            if (parsed.publishDate) {
              parsed.publishDate = "NORMALIZED_DATE";
            }
            if (parsed.updateDate) {
              parsed.updateDate = "NORMALIZED_DATE";
            }
            if (parsed.variants && Array.isArray(parsed.variants)) {
              parsed.variants = parsed.variants.map((variant: any) => {
                if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
                if (variant.publishDate)
                  variant.publishDate = "NORMALIZED_DATE";
                if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
                return variant;
              });
            }
            text = JSON.stringify(parsed);
          } catch {}
          return {
            ...item,
            text,
          };
        } else {
          // For list responses
          const parsed = JSON.parse(item.text);
          if (Array.isArray(parsed)) {
            // Handle ancestors API response
            return {
              ...item,
              text: JSON.stringify(parsed.map(normalizeItem)),
            };
          }
          // Handle other list responses
          if (parsed.items) {
            parsed.items = parsed.items.map(normalizeItem);
          }
          if (parsed.variants && Array.isArray(parsed.variants)) {
            parsed.variants = parsed.variants.map((variant: any) => {
              if (variant.createDate) variant.createDate = "NORMALIZED_DATE";
              if (variant.publishDate) variant.publishDate = "NORMALIZED_DATE";
              if (variant.updateDate) variant.updateDate = "NORMALIZED_DATE";
              return variant;
            });
          }
          return {
            ...item,
            text: JSON.stringify(parsed),
          };
        }
      }
      return item;
    }),
  };
}

export function normalizeErrorResponse(result: CallToolResult): CallToolResult {
  if (
    Array.isArray(result.content) &&
    result.content[0]?.text &&
    typeof result.content[0].text === "string"
  ) {
    // Replace any traceId in the text with a normalized version
    result.content[0].text = result.content[0].text.replace(
      /00-[0-9a-f]{32}-[0-9a-f]{16}-00/g,
      "normalized-trace-id"
    );
  }
  return result;
}
