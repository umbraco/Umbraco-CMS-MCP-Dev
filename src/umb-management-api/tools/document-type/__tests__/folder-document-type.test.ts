import { DocumentTypeTestHelper } from "./helpers/document-type-test-helper.js";
import CreateDocumentTypeFolderTool from "../folders/post/create-folder.js";
import DeleteDocumentTypeFolderTool from "../folders/delete/delete-folder.js";
import UpdateDocumentTypeFolderTool from "../folders/put/update-folder.js";
import { createSnapshotResult } from "@/test-helpers/create-snapshot-result.js";
import { jest } from "@jest/globals";
import { DocumentTypeFolderBuilder } from "./helpers/document-type-folder-builder.js";
import { BLANK_UUID } from "@/constants/constants.js";

describe("document-type-folder", () => {
  const TEST_FOLDER_NAME = "_Test DocumentType Folder";
  const TEST_PARENT_FOLDER_NAME = "_Test Parent Folder";
  const UPDATE_FOLDER_NAME = "_Update Folder Name";
  const UPDATED_FOLDER_NAME = "_Updated Folder Name";
  let originalConsoleError: typeof console.error;

  beforeEach(() => {
    originalConsoleError = console.error;
    console.error = jest.fn();
  });

  afterEach(async () => {
    console.error = originalConsoleError;
    await DocumentTypeTestHelper.cleanup(TEST_FOLDER_NAME);
    await DocumentTypeTestHelper.cleanup(TEST_PARENT_FOLDER_NAME);
  });

  describe("create", () => {
    it("should create a folder", async () => {
      const result = await CreateDocumentTypeFolderTool().handler(
        {
          name: TEST_FOLDER_NAME,
        },
        { signal: new AbortController().signal }
      );

      expect(createSnapshotResult(result)).toMatchSnapshot();

      // Verify folder exists
      const found = await DocumentTypeTestHelper.findDocumentType(
        TEST_FOLDER_NAME
      );
      expect(found).toBeDefined();
      expect(found!.isFolder).toBe(true);
    });

    it("should create a folder with parent", async () => {
      // Create parent folder using builder
      const parentBuilder = await new DocumentTypeFolderBuilder(
        TEST_PARENT_FOLDER_NAME
      ).create();
      expect(parentBuilder).toBeDefined();

      const result = await CreateDocumentTypeFolderTool().handler(
        {
          name: TEST_FOLDER_NAME,
          parent: { id: parentBuilder.getId() },
        },
        { signal: new AbortController().signal }
      );

      expect(createSnapshotResult(result)).toMatchSnapshot();

      // Verify folder exists under parent
      const found = await DocumentTypeTestHelper.findDocumentType(
        TEST_FOLDER_NAME
      );
      expect(found).toBeDefined();
      expect(found!.isFolder).toBe(true);
    });
  });

  describe("update", () => {
    it("should update a folder name", async () => {
      // Create folder to update using builder
      const builder = await new DocumentTypeFolderBuilder(
        UPDATE_FOLDER_NAME
      ).create();
      expect(builder).toBeDefined();

      const result = await UpdateDocumentTypeFolderTool().handler(
        {
          id: builder.getId(),
          data: {
            name: UPDATED_FOLDER_NAME,
          },
        },
        { signal: new AbortController().signal }
      );

      expect(result).toMatchSnapshot();

      // Verify folder was updated
      const found = await DocumentTypeTestHelper.findDocumentType(
        UPDATED_FOLDER_NAME
      );
      expect(found).toBeDefined();
      expect(found!.name).toBe(UPDATED_FOLDER_NAME);
      await DocumentTypeTestHelper.cleanup(UPDATED_FOLDER_NAME);
    });

    it("should handle non-existent folder", async () => {
      const result = await UpdateDocumentTypeFolderTool().handler(
        {
          id: BLANK_UUID,
          data: {
            name: UPDATED_FOLDER_NAME,
          },
        },
        { signal: new AbortController().signal }
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe("delete", () => {
    it("should delete a folder", async () => {
      // Create folder to delete using builder
      const builder = await new DocumentTypeFolderBuilder(
        TEST_FOLDER_NAME
      ).create();
      expect(builder).toBeDefined();

      const result = await DeleteDocumentTypeFolderTool().handler(
        {
          id: builder.getId(),
        },
        { signal: new AbortController().signal }
      );

      expect(createSnapshotResult(result)).toMatchSnapshot();

      // Verify folder was deleted
      const found = await DocumentTypeTestHelper.findDocumentType(
        TEST_FOLDER_NAME
      );
      expect(found).toBeUndefined();
    });

    it("should handle non-existent folder", async () => {
      const result = await DeleteDocumentTypeFolderTool().handler(
        {
          id: BLANK_UUID,
        },
        { signal: new AbortController().signal }
      );

      expect(result).toMatchSnapshot();
    });
  });
});
