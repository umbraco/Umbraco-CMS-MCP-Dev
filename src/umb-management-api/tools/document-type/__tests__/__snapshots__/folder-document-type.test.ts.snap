// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document-type-folder create should create a folder 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;

exports[`document-type-folder create should create a folder with parent 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;

exports[`document-type-folder delete should delete a folder 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;

exports[`document-type-folder delete should handle non-existent folder 1`] = `
{
  "content": [
    {
      "text": "Error using delete-document-type-folder:
{
  "message": "Request failed with status code 404",
  "response": {
    "type": "Error",
    "title": "The folder could not be found",
    "status": 404,
    "operationStatus": "NotFound"
  }
}",
      "type": "text",
    },
  ],
}
`;

exports[`document-type-folder update should handle non-existent folder 1`] = `
{
  "content": [
    {
      "text": "Error using update-document-type-folder:
{
  "message": "Request failed with status code 404",
  "response": {
    "type": "Error",
    "title": "The folder could not be found",
    "status": 404,
    "operationStatus": "NotFound"
  }
}",
      "type": "text",
    },
  ],
}
`;

exports[`document-type-folder update should update a folder name 1`] = `
{
  "content": [
    {
      "text": """",
      "type": "text",
    },
  ],
}
`;
